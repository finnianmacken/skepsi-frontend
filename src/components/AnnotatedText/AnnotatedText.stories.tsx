import React from "react"
import { Meta } from '@storybook/react'
import AnnotatedText from "../AnnotatedText/AnnotatedText"
import { Set } from "immutable"
import { Annotation } from "../types"

import "../../App.css"

export default {
    title: 'DocumentViewer/AnnotatedText',
    component: AnnotatedText,
} as Meta

export const Example = () => (
    <AnnotatedText
        annotations={Set([
            {start: 5, stop: 18, data: {props: {style: {fontWeight: 'bold'}}}},
            {start: 30, stop: 40},
            {start: 38, stop: 50}
        ]).map(a => Annotation(a))}>
        the first highlight asks for bold formatting, the second two overlap
    </AnnotatedText>
)
