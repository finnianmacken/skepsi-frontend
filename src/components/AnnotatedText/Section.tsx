import React from "react"
import {Set} from "immutable"
import {AnnotationType} from "../types"

const highlightStyle = (annotations: Set<AnnotationType>): object => {
    const depth = annotations.size
    const colors = ["transparent", "yellow", "orange", "red", "purple"]
    return {backgroundColor: colors[Math.min(depth, colors.length)]}
}

type SectionType = {
    annotations: Set<AnnotationType>
    children: string
}

const Section: React.FC<SectionType> = (props) => {
    const annotationProps = props.annotations
        .map(a => {
            try {
                return a.data.sectionProps
            } catch (e) {
                return {}
            }
        })
        .reduce((a, b) => ({...a, ...b}), {style: {}})

    return (
        <span
            {...annotationProps}
            style={{
                ...annotationProps.style,
                ...highlightStyle(props.annotations),
                position: "relative",
            }}
            className={"Section"}>
            {props.children}
        </span>
    )
}

export default Section
