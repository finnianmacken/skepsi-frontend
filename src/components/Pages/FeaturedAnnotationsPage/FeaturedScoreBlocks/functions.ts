// type variables impossible in tsx?
export const optionizeArray = <T>(vals: T[]) => vals.map(x => ({key: x, text: x, value: x}))
