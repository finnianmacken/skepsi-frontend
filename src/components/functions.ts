import React from "react"
import {Collection, List, RecordOf, Set} from "immutable"
import renderer from "react-test-renderer"
import {ContentNodeType} from "./types"
import curry from "just-curry-it"

/*
Whether the list is ascending monotonically (a > b)
 */
export const isAscending = (arr: Array<number>) => arr.every(
    (x, i) => i === 0 || x > arr[i - 1]
)

export const convolve = (<ItemType, ReturnType, StateType>(fn: (item: ItemType, state: StateType) => [ReturnType, StateType], initial: StateType, items: List<ItemType>) => {
    let currentState = initial
    return items.map((item: ItemType) => {
        const [r, s] = fn(item, currentState)
        currentState = s
        return r
    })
})

export const printImmutable = (c: Collection<any, any> | RecordOf<any>, multiline?: boolean) => {
    if (multiline) {
        console.log(JSON.stringify(c.toJS(), null, 4))
    } else {
        console.log(JSON.stringify(c.toJS()))
    }
}

/*
Snapshot testing for stories files. Usage:

```
const stories = require("./FooComponent.stories");
testStories(stories);
```
 */
export const testStories = (stories: any) => {
    const sObject = Object.create(stories)
    const toTest = []

    for (let name in stories) {
        if (name !== "default"){
            toTest.push(sObject[name]) // raises linter warning
        }
    }

    test.each(toTest)(
        "render",
        c => expect(renderer.create(React.createElement(c)).toJSON()).toMatchSnapshot()
    )
}

export const nodeIdToPrettyId = (nodeId: number) => "pos-" + nodeId
export const nodePrettyId = (node: ContentNodeType) => nodeIdToPrettyId(node._id)

export const scrollToContentNode = (node: ContentNodeType) => {
    document.getElementById(nodePrettyId(node))?.scrollIntoView({behavior: "smooth", block: "center"})
}

export const spy = <T>(thing: T): T => {
    console.log(thing);
    return thing
}

export const spyWith = curry((fn: Function, thing: any) => spy(fn(thing)))

export const formatDate = (date: Date) => {
    let d = new Date(date),
        month = '' + (d.getMonth() + 1),
        day = '' + d.getDate(),
        year = d.getFullYear();

    if (month.length < 2)
        month = '0' + month;
    if (day.length < 2)
        day = '0' + day;

    return [year, month, day].join('-');
}

export const maxOfIterable = <T>(lst: List<T> | Set<T> | Array<T>, val: (x: T) => number): T | undefined => {
    let best: T | undefined = undefined
    let bestVal = -Infinity

    lst.forEach(item => {
        const v = val(item)
        if (v > bestVal) {
            bestVal = v
            best = item
        }
    })
    return best
}

export const range = (start: number, stop?: number, step?: number) => {
    step = step || 1
    if (stop === undefined){
        stop = start
        start = 0
    }

    if (Math.sign(start - stop) === Math.sign(step)){
        return []
    }

    const arr = []
    while (Math.sign(start - stop) === -Math.sign(step)){
        arr.push(start)
        start += step
    }
    return arr
}

export const zip = <T, U>(a1: T[], a2: U[]): [T, U][] => (
    range(Math.min(a1.length, a2.length)).map(i => [a1[i], a2[i]])
)
