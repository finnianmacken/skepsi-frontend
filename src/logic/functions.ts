/*

A bunch of general purpose functions that work on the the types etc. specific to this project

 */

import {List, Set} from "immutable"
import {isAscending} from "../utility/functions"
import {ContentNodeType} from "./contentNode"
import {AnnotationType} from "./annotation"

export const nodeIdToPrettyId = (nodeId: number) => "pos-" + nodeId
export const nodePrettyId = (node: ContentNodeType) => nodeIdToPrettyId(node._id)

export const scrollToContentNode = (node: ContentNodeType) => {
    document.getElementById(nodePrettyId(node))?.scrollIntoView({behavior: "smooth", block: "center"})
}

// whether the annotation overlaps with the region designated by start and stop
export const isAnnotationSectionOverlap = (start: number, stop: number, a: AnnotationType) => (
    !(isAscending([a.start, a.stop, start, stop]) || isAscending([start, stop, a.start, a.stop]))
)

/*
Translates the annotation into the reference frame defined by start and stop
 */
export const translateAnnotation = (start: number, stop: number, a: AnnotationType) => {
    if (!isAnnotationSectionOverlap(start, stop, a)){
        throw Error(`no overlap between annotation ${[a.start, a.stop]} and section ${[start, stop]}`)
    }
    return a.merge({
        start: Math.max(0, a.start - start),
        stop: Math.min(stop - start, a.stop - start),
    })
}

/*
Takes a root node a returns a list of all the nodes in the underlying tree
 */
export const nodesInNode = (node: ContentNodeType): List<ContentNodeType> => {
    if (node.content === undefined || typeof node.content === "string"){
        return List([node])
    } else if (!List.isList(node.content)) {
        return List([node, node.content])
    } else {
        return List([node]).concat(
            node.content
                .map(n => nodesInNode(n))
                .reduce((a, b) => a.concat(b), List<ContentNodeType>())
        )
    }
}

/*
recursive map for document node functor
 */
export const nodeMap = (node: ContentNodeType, fn: (n: ContentNodeType) => ContentNodeType): ContentNodeType => {

    let content
    if (node.content === undefined || typeof node.content === "string"){
        content = node.content
    } else if (List.isList(node.content)) {
        content = node.content.map(n => nodeMap(n, fn))
    } else {
        content = nodeMap(node.content, fn)
    }

    return fn(node.merge({content: content}))
}

export const addSingleAnnotation = (root: ContentNodeType, annotation: AnnotationType): [ContentNodeType, AnnotationType] => {
    root = nodeMap(root, (node) => {
        if (typeof node.content === "string"){
            const start = node.startIndex
            const stop = node.startIndex + node.content.length

            if (isAnnotationSectionOverlap(start, stop, annotation)){
                node = node.merge({props: {
                        ...node.props,
                        annotations: (node.props.annotations || Set<AnnotationType>()).add(
                            translateAnnotation(start, stop, annotation)
                        )
                    }})
                annotation = annotation.merge({_node_id: annotation._node_id.push(node._id)})
            }
        }

        // returns the same node object, allowing memoization
        return node
    })

    return [root, annotation]
}
