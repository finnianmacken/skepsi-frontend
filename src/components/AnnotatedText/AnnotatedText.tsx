import React from "react"
import {List, Set} from "immutable"
import {assignAnnotations} from "../../logic/processing"
import {AnnotationType} from "../../logic/annotation"
import Section from "./Section"

const breakUpString = (content: string, annotations: Set<AnnotationType>) => {
    const breakpoints = List(Set(               // kills duplicates
        (annotations
            .map(a => List.of(a.start, a.stop))
            .flatten() as List<number>)             // one list of start and stop positions for every annotation
            .concat([0, content.length])
    )).sort()

    return breakpoints
        .pop().zip(breakpoints.shift()) // zip with offset self
        .map(([start, stop]) => content.substring(start, stop))
}

type AnnotatedTextType = {
    children: string
    annotations?: Set<AnnotationType>
    withProps?: object
    setActiveAnnotationId?: (val: number | ((id: number) => number)) => void
}

const AnnotatedText: React.FC<AnnotatedTextType> = (props) => {
    const annotations: Set<AnnotationType> = props.annotations || Set()
    return (
        <span {...(props.withProps || {})} className={"AnnotatedText"}>
            {assignAnnotations(x => x.length, annotations, breakUpString(props.children, annotations))
                .map(([substring, substringAnnotations], i) =>
                    <Section key={i} annotations={substringAnnotations} setActiveAnnotationId={props.setActiveAnnotationId}>
                        {substring}
                    </Section>)}
        </span>
    )
}

export default AnnotatedText
