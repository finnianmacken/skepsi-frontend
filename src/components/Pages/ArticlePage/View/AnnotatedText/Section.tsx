import React from "react"
import {List, Set, Map} from "immutable"
import {AnnotationType} from "logic/annotation"

const highlightStyle = (annotations: Set<AnnotationType>): object => {
    const depth = annotations.size

    const colors = ["transparent", "#e2bc92", "#bd997e", "#9b7768", "#725a57", "#523b3d"]
    const bottomColors = ["transparent", "#e2bc92", "#bd997e", "#9b7768", "#725a57", "#523b3d"]

    return {
        backgroundColor: colors[Math.min(depth, colors.length)],
        borderBottom: depth > 0 ? "2px solid " + bottomColors[Math.min(depth, bottomColors.length)]: "",
    }
}

type SectionType = {
    annotations: Set<AnnotationType>
    children: string
    setActiveAnnotationId?: (val: number | ((id: number) => number)) => void
}

const Section: React.FC<SectionType> = (props) => {
    const annotationProps = props.annotations
        .map(a => {
            try {
                return a.data.sectionProps
            } catch (e) {
                return {}
            }
        })
        .reduce((a, b) => ({...a, ...b}), {style: {}})

    const ids = List(props.annotations.map(a => a._id)).sort()
    const idMap = Map(ids.zip(ids.shift().push(ids.first())))

    const onClick = () => {
        if (props.setActiveAnnotationId){
            props.setActiveAnnotationId((id: number) => idMap.get(id, ids.get(0, NaN)))
        }
    }

    return (
        <span
            {...annotationProps}
            onMouseDown={onClick} // for some reason onclick isn't working
            style={{
                ...annotationProps.style,
                ...highlightStyle(props.annotations),
                position: "relative",
            }}
            className={"Section"}>
            {props.children}
        </span>
    )
}

export default Section
