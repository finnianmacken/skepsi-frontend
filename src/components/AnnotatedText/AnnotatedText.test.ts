import {testStories} from "../functions"
import * as stories from "./AnnotatedText.stories"
testStories(stories)
