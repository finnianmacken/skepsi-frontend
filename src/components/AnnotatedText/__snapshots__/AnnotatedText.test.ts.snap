// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<span
  className="AnnotatedText"
>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "transparent",
        "position": "relative",
      }
    }
  >
    the f
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "yellow",
        "position": "relative",
      }
    }
  >
    irst highligh
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "transparent",
        "position": "relative",
      }
    }
  >
    t asks for b
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "yellow",
        "position": "relative",
      }
    }
  >
    old form
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "yellow",
        "position": "relative",
      }
    }
  >
    at
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "yellow",
        "position": "relative",
      }
    }
  >
    ting, the 
  </span>
  <span
    className="Section"
    style={
      Object {
        "backgroundColor": "transparent",
        "position": "relative",
      }
    }
  >
    second two overlap
  </span>
</span>
`;
